src/main/java/com/tabnine/binary/fetch/BootstrapperSupport.java

```diff
@@ -0,0 +1,47 @@
+package com.tabnine.binary.fetch;
+
+import com.intellij.util.text.SemVer;
+import com.tabnine.general.StaticConfig;
+
+import java.util.Optional;
+import java.util.prefs.Preferences;
+
+// bootstrapper support class
+// once out of beta/alpha all of this code can replace the existing fetchBinary()
+public class BootstrapperSupport {
+    static Optional<BinaryVersion> bootstrapVersion(LocalBinaryVersions localBinaryVersions, BinaryRemoteSource binaryRemoteSource, BundleDownloader bundleDownloader) {
+        Optional<BinaryVersion> localBootstrapVersion = locateLocalBootstrapSupportedVersion(localBinaryVersions);
+        return localBootstrapVersion.isPresent() ? localBootstrapVersion : downloadRemoteVersion(binaryRemoteSource, bundleDownloader);
+    }
+
+    private static final String BOOTSTRAPPED_VERSION_KEY = "bootstrapped version";
+
+    private static Preferences getPrefs() {
+        return Preferences.userNodeForPackage(BootstrapperSupport.class);
+    }
+
+    private static BinaryVersion savePreferredBootstrapVersion(BinaryVersion version) {
+        getPrefs().put(BOOTSTRAPPED_VERSION_KEY, version.getVersion());
+        return version;
+    }
+
+    private static Optional<SemVer> minimalBootstrappedVersion() {
+        return Optional.ofNullable(getPrefs().get(BOOTSTRAPPED_VERSION_KEY, null)).map(SemVer::parseFromText);
+    }
+
+    private static Optional<BinaryVersion> locateLocalBootstrapSupportedVersion(LocalBinaryVersions localBinaryVersions) {
+        return minimalBootstrappedVersion().flatMap(
```

Comment:
shouldn't we always take the most recent one? Maybe I'm mssing something, but `findFirst` indicates that we can select pretty much any of them, no?

Reply:
it will

Reply:
`.filter(v -> SemVer.parseFromText(v.getVersion()).compareTo(version) >= 0)` and then taking the first (which is >= version)

Reply:
:+1:

Reply:
Consider renaming the method or something, because at least for me, it was not obvious form the code